src/Microsoft.DotNet.Build.Tasks/PackageFiles/versioning.targets

```diff
@@ -158,65 +158,74 @@
   <Target Name="CreateOrUpdateCurrentVersionFile" Condition="'$(SkipVersionGeneration)'!='true'" DependsOnTargets="GenerateCurrentVersion">
-    <Error Condition="'$(OverrideBuildNumberMajor)'==''" Text="Error when generating current version." />
+    <!-- Error out if GeneratedBuildNumberMajor is empty since GenerateCurrentVersion should have calculated it already. -->
+    <Error Condition="'$(GeneratedBuildNumberMajor)'==''" Text="Error when generating current version." />
 
     <ItemGroup>
       <CurrentVersionLines Include="%3C%3Fxml version=%221.0%22 encoding=%22utf-8%22%3F%3E" />
-      <CurrentVersionLines Include="%3C!-- This is a generated file. To re-generate it, run build.cmd --%3E" />
+      <CurrentVersionLines Include="%3C!-- This is a generated file. --%3E" />
       <CurrentVersionLines Include="%3CProject xmlns=%22http://schemas.microsoft.com/developer/msbuild/2003%22%3E" />
       <CurrentVersionLines Include="%3CPropertyGroup%3E" />
-      <CurrentVersionLines Include="%3CBuildNumberMajor Condition=%22%27%24(BuildNumberMajor)%27==%27%27%22%3E$(OverrideBuildNumberMajor)%3C/BuildNumberMajor%3E" />
+      <CurrentVersionLines Include="%3CBuildNumberMajor Condition=%22%27%24(BuildNumberMajor)%27==%27%27%22%3E$(GeneratedBuildNumberMajor)%3C/BuildNumberMajor%3E" />
       <CurrentVersionLines Include="%3C/PropertyGroup%3E" />
       <CurrentVersionLines Include="%3C/Project%3E" />
       <BuildVersionFileItem Include="$(BuildVersionFile)" />
     </ItemGroup>
-
+    
     <PropertyGroup>
-      <BuildVersionSemaphoreFile>%(BuildVersionFileItem.RootDir)%(BuildVersionFileItem.Directory)semaphore.txt</BuildVersionSemaphoreFile>
-      <ShouldLockFileWrite Condition="!Exists('$(BuildVersionSemaphoreFile)')">true</ShouldLockFileWrite>
+      <BuildVersionFilePath>%(BuildVersionFileItem.RootDir)%(BuildVersionFileItem.Directory)</BuildVersionFilePath>
     </PropertyGroup>
+    
 
-    <MakeDir Condition="!Exists('%(BuildVersionFileItem.RootDir)%(BuildVersionFileItem.Directory)')" Directories="%(BuildVersionFileItem.RootDir)%(BuildVersionFileItem.Directory)" />
-
-    <Touch AlwaysCreate="true" Files="$(BuildVersionSemaphoreFile)" Condition="'$(ShouldLockFileWrite)'=='true'" />
-
+    <!-- Since by default the file will get dropped at the obj dir, make sure that the dir is created already or else WriteLinesToFile will error. -->
+    <MakeDir Condition="!Exists('$(BuildVersionFilePath)')" Directories="$(BuildVersionFilePath)" />
+      
     <WriteLinesToFile
       ContinueOnError="WarnAndContinue"
-      Condition="'$(ShouldLockFileWrite)'=='true'"
+      Condition="!Exists('$(BuildVersionFile)')"
       File="$(BuildVersionFile)"
       Lines="@(CurrentVersionLines)"
       Overwrite="true" />
+    
+    <!-- Delete old BuildVersion.props files -->
+    <ItemGroup>
+      <OldBuildVersionFiles Include="$(BuildVersionFilePath)*.props" Exclude="$(BuildVersionFilePath)%(BuildVersionFileItem.Filename).props" />
```

Comment:
To be safe you should use `$(BuildVersionFilePath)BuildVersions-*.props`. Also thinking about this a minute perhaps doing the delete before you write the new file would eliminate the need to exclude it.